public/app/features/alerting/unified/rule-list/GroupedView.tsx

```diff
@@ -0,0 +1,280 @@
+import { css } from '@emotion/css';
+import { PropsWithChildren, ReactNode, useMemo } from 'react';
+import Skeleton from 'react-loading-skeleton';
+
+import { GrafanaTheme2 } from '@grafana/data';
+import { Button, Dropdown, Icon, IconButton, LinkButton, Menu, Stack, Text, useStyles2 } from '@grafana/ui';
+import { Trans } from 'app/core/internationalization';
+import { DataSourceNamespaceIdentifier, DataSourceRuleGroupIdentifier, RuleGroup } from 'app/types/unified-alerting';
+import { RulesSourceApplication } from 'app/types/unified-alerting-dto';
+
+import { featureDiscoveryApi } from '../api/featureDiscoveryApi';
+import { Spacer } from '../components/Spacer';
+import { WithReturnButton } from '../components/WithReturnButton';
+import { getDatasourceAPIUid, getExternalRulesSources } from '../utils/datasource';
+import { hashRule } from '../utils/rule-id';
+
+import { AlertRuleLoader } from './AlertRuleLoader';
+import { ListGroup } from './components/ListGroup';
+import { ListSection } from './components/ListSection';
+import { DataSourceIcon } from './components/Namespace';
+import { LoadingIndicator } from './components/RuleGroup';
+import { usePaginatedPrometheusRuleNamespaces } from './hooks/usePaginatedPrometheusRuleNamespaces';
+
+const { useDiscoverDsFeaturesQuery } = featureDiscoveryApi;
+const GROUP_PAGE_SIZE = 40;
+
+export function GroupedView() {
+  const externalRuleSources = useMemo(() => getExternalRulesSources(), []);
```

Comment:
why do we need this memoization?

Reply:
Just performance reasons, to avoid calculating external rules sources all the time